docs/docs/adrs/adr-011-improving-test-confidence.md

```diff
@@ -0,0 +1,208 @@
+- non-determinism tests
+- upgrade tests
+- state-breaking changes
+
+With CometMock, the **matrix test** approach can also be used. The image below illustrates a CometMock setup that can be used to discover non-deterministic behavior and state-breaking changes.
+![e2e matrix tests](../../figures/cometmock_matrix_test.png)
+
+This table explains which versions are tested against each other for the same set of test traces:
+* ✅ marks a passing test
+* ❌ marks a failing test
+
+| **SCENARIO**                    | **start chain** | **add key** | **delegate** | **undelegate** | **redelegate** | **downtime** | **equivocation** | **stop chain** |
+|---------------------------------|-----------------|-------------|--------------|----------------|----------------|--------------|------------------|----------------|
+| **v3 provi + v3 consu**         | ✅               | ✅           | ✅            | ✅              | ✅              | ✅            | ✅                | ✅              |
+| **main provi + main consu**     | ✅               | ✅           | ✅            | ✅              | ✅              | ✅            | ✅                | ✅              |
+| **commit provi + commit consu** | ✅               | ❌           | ✅            | ❌              | ✅              | ✅            | ❌                | ❌              |
+
+
+
+Briefly; multiple versions of the application are run against the same CometMock instance and any deviations in app behavior would result in `app hash` errors (the apps would be in different states after performing the same set of actions).
+
+#### 3. Introduce innovative testing approaches
+
+When discussing e2e testing, some very important patterns emerge - especially if test traces are used instead of ad-hoc tests written by hand.
+
+We see a unique opportunity to clearly identify concerns and modularize the testing architecture. 
+
+The e2e testing frameworks can be split into a **pipeline consisting of 3 parts: model, driver and harness**.
+
+##### Model
+
+Model is the part of the system that can emulate the behavior of the system under test.
+Ideally, it is very close to the specification and is written in a specification language such as quint, TLA+ or similar.
+One of the purposes of the model is that it can be used to generate test traces. 
+
+
+##### Driver
+
+The purpose of the driver is to accept test traces (generated by the model or written by hand), process them and provide inputs to the next part of the pipeline.
+
+Basically, the driver sits between the model and the actual infrastructure on which the test traces are being executed on.
+
+##### Harness
+
+Harness is the infrastructure layer of the pipeline that accepts inputs from the driver.
+
+There can be multiple harnesses as long as they can perform 3 things:
+* bootstrap a test execution environment (local, docker, k8s…)
+* accept inputs from drivers
+* perform the action specified by the driver
+* report results after performing actions
+
+
+## Consequences
```

Comment:
Might be worth keeping in mind that the procedure outlined in the adr is not all-or-nothing - CometMock, Matrix Tests, MBT, ... are mostly independent and we can decide to do one of those and not the other without issue.